master/internal/core_checkpoint.go

```diff
@@ -0,0 +1,470 @@
+				", but the checkpoint's storage type is %s", storageConfig2Str(storage)))
+	}
+}
+
+// It is assumed that a http status code is not sent until the first write to w.
+func buildWriterPipeline(w io.Writer, mimeType string) (archiveWriter, error) {
+	// DelayWriter delays the first write until we have successfully downloaded
+	// some bytes and are more confident that the download will succeed.
+	dw := newDelayWriter(w, 16*1024)
+	closers := []io.Closer{dw}
+	switch mimeType {
+	case MIMEApplicationGZip:
+		gz := gzip.NewWriter(dw)
+		closers = append(closers, gz)
+
+		tw := tar.NewWriter(gz)
+		closers = append(closers, tw)
+
+		return &tarArchiveWriter{archiveClosers{closers}, tw}, nil
+
+	case MIMEApplicationZip:
+		zw := zip.NewWriter(dw)
+		closers = append(closers, zw)
+
+		return &zipArchiveWriter{archiveClosers{closers}, zw, nil}, nil
+
+	default:
+		return nil, fmt.Errorf(
+			"MIME type must be %s or %s but got %s",
+			MIMEApplicationGZip, MIMEApplicationZip, mimeType)
+	}
+}
+
+func (m *Master) getCheckpointStorageConfig(id uuid.UUID) (
+	*expconf.CheckpointStorageConfig, error) {
+	checkpoint, err := m.db.CheckpointByUUID(id)
+	if err != nil || checkpoint == nil {
+		return nil, err
+	}
+
+	bytes, err := json.Marshal(checkpoint.CheckpointTrainingMetadata.ExperimentConfig)
+	if err != nil {
+		return nil, err
+	}
+	spew.Dump(bytes)
+	spew.Dump(checkpoint.CheckpointTrainingMetadata.ExperimentConfig)
+
+	legacyConfig, err := expconf.ParseLegacyConfigJSON(bytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return ptrs.Ptr(legacyConfig.CheckpointStorage()), nil
+}
+
+func (m *Master) getCheckpoint(c echo.Context, mimeType string) error {
+	args := struct {
+		CheckpointUUID string `path:"checkpoint_uuid"`
+	}{}
+	if err := api.BindArgs(&args, c); err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest,
+			"invalid checkpoint_uuid: "+err.Error())
+	}
+
+	checkpointUUID, err := uuid.Parse(args.CheckpointUUID)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest,
+			fmt.Sprintf("unable to parse checkpoint UUID %s: %s",
+				args.CheckpointUUID, err))
+	}
+
+	// Assume a checkpoint always has experiment configs
+	storageConfig, err := m.getCheckpointStorageConfig(checkpointUUID)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusInternalServerError,
+			fmt.Sprintf("unable to retrieve experiment config for checkpoint %s: %s",
+				args.CheckpointUUID, err.Error()))
+	}
+	if storageConfig == nil {
+		return echo.NewHTTPError(http.StatusNotFound,
+			fmt.Sprintf("checkpoint %s does not exist", args.CheckpointUUID))
+	}
+
+	c.Response().Header().Set(echo.HeaderContentType, mimeType)
+	writerPipe, err := buildWriterPipeline(c.Response(), mimeType)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest, err.Error())
+	}
+
+	downloader, err := newDownloader(storageConfig, writerPipe, args.CheckpointUUID)
+	err = downloader.download(c.Request().Context())
+	if err != nil {
```

Comment:
this error handling is kinda funky, maybe explicitly handle both errors?

Reply:
Yeah, I missed the one returned by `newDownloader()` if that's what you mean.